rfc/0019-hashed-mojmap.md

```diff
@@ -0,0 +1,97 @@
+# Hashed Mojmap
+
+## Summary
+
+This RFC proposes replacing intermediary with a hashed version of Mojmap. Everywhere where intermediary is currently used, hashed Mojmap will be used instead.
+
+## Motivation
+
+- One of the biggest, if not the biggest, reliance of Quilt on upstream Fabric is intermediary. For as long as Fabric keeps updating to new Minecraft versions, we will be behind them on updating because of this limitation. It is also not a good thing for a hard fork to have a reliance on upstream.
+
+- One of the primary goals of intermediary is to make sure that mods don't have to recompile between versions. However, when mods make the switch from Fabric to Quilt, they will have to recompile anyway. This gives us a unique opportunity to make breaking changes to the format of intermediary without breaking mods.
+
+- Intermediary was a good solution at the time of its inception. Since then, Mojang has provided us with their official mappings, with permission to use them as long as we don't redistribute them. Having to match intermediary between versions introduces a high maintenance cost for every snapshot and potentially also limits the bus factor of the whole project. The majority of the time updating Fabric to snapshots is typically spent matching.
+
+
+## Explanation
+
+Hashed Mojmap will produce base-62 hashes of information taken from Mojmap to produce mixed-case alphanumeric names like the following:
+
+- Classes: `Cls_hE89sr` (6 digits)
+- Fields: `fd_p9TXc40` (7 digits)
+- Methods: `md_fjH9ac8` (7 digits)
+
+To produce the hashes, the Mojmap names are processed as follows:
+
+1. Classes are converted from dotty format (`net.minecraft.a.Foo$Bar`) to slashy format (`net/minecraft/a/Foo$Bar`).
+1. Overriding methods in subclasses are discarded if present.
+1. Package names are removed from classes whose simple names are unique. That is, `net/minecraft/a/Bar` gets shortened to `Bar` if there is no other class named `Bar`, but if there are classes `net/minecraft/a/Foo` and `net/minecraft/b/Foo` that exist, then neither of them will be shortened. This includes references to these classes in method descriptors.
+1. Method descriptors are replaced with an empty string for methods whose name is unique within the class that the method is declared.
+1. Field names are converted to `fd_$x` where `$x` is the last 7 digits of the base-62 representation of the SHA-1 hash of the string `f$declaring_class.$field_name`.
+1. Method names are converted to `md_$x` where `$x` is the last 7 digits of the base-62 representation of the SHA-1 hash of the string `m$declaring_class.$method_name$descriptor`.
+1. Class names are converted to `Cls_$x` where `$x` is the last 6 digits of the base-62 representation of the SHA-1 hash of the class name.
```

Comment:
Why SHA-1? Of course cryptographic security/easier collisions are not an issue, but wouldn't a newer standard like SHA-256 (SHA-3 might not be as widely supported) be the better choice for a new proposal like this?

Reply:
SHA-1 is a little faster than SHA-2, which multiplies up when you're hashing tens of thousands of strings individually. Still probably won't matter that much but if there's nothing else in it it's always nice to go for speed.

Reply:
Apparently I was wrong, and SHA-256 is faster than SHA-1. I will change this.